csrc/fusion_segmenter.cpp

```diff
@@ -2059,54 +2059,92 @@ bool SegmentCandidateFinder::hasSegmentHints(Fusion* fusion) {
   return false;
 }
 
-void SegmentCandidateFinder::resetTraversal() {
+void SegmentCandidateFinder::resetLevels() {
+  FUSER_PERF_SCOPE("SegmentCandidateFinder::resetLevels");
+
+  std::deque<SegmentedGroup*> to_visit;
+  std::vector<SegmentedGroup*> next_to_visit;
+
+  // Special initialization for
+  // https://github.com/Lightning-AI/lightning-thunder/issues/1490 and
+  // https://github.com/NVIDIA/Fuser/issues/3510 (same issue).
+  //
+  // The issue is if we have a DAG that's a long chain that keeps adding inputs.
+  // e.g.
+  // tv3 = inp0 + inp1
+  // tv4 = tv3
+  // tv6 = tv4 + inp5
+  // tv7 = tv6
+  // tv9 = tv7 + inp8
+  // ...
+  //
+  // If we simply start from all inputs, then we will add all expressions to the
+  // to_visit list. We will check the first expression, process it, add the next
+  // expression to the end of the to_visit list. Then check all other expression
+  // which cannot be processed, until finally getting back to the second
+  // expression, and process it.
+  //
+  // The solution to this is to start from expressions where all producer edges
+  // are inputs.
+
+  // Mark all input groups as viisted and mark their level as 0
+  // Initialize all other groups as not visited and level as 0
   for (auto group : groups()) {
-    // Start traversal at input groups
+    group->level_ = 0;
     if (group->producer_edges.empty()) {
-      to_visit_.push_back(group);
+      group->visited_ = true;
+    } else {
+      group->visited_ = false;
+      if (std::all_of(
+              group->producer_edges.begin(),
+              group->producer_edges.end(),
+              [&](SegmentedEdge* edge) {
+                return edge->from->producer_edges.empty();
+              })) {
+        to_visit.push_back(group);
+      }
     }
-    group->visited_ = false;
-    group->level_ = 0;
   }
-}
 
-void SegmentCandidateFinder::resetLevels() {
-  while (!to_visit_.empty()) {
-    auto visit = to_visit_.front();
-    to_visit_.pop_front();
-
-    // All inputs processed?
-    bool ready = true;
-    if (!visit->producer_edges.empty()) {
-      ready = std::all_of(
-          visit->producer_edges.begin(),
-          visit->producer_edges.end(),
-          [&](SegmentedEdge* dep) { return dep->from->visited_; });
-    }
-
-    if (!ready) {
-      // In case traversal doesn't complete because there's an error in the
-      // DAG topology.
-      next_to_visit_.push_back(visit);
+  while (!to_visit.empty()) {
+    auto visiting = to_visit.front();
+    to_visit.pop_front();
+
+    if (visiting->visited_) {
       continue;
     }
 
-    visit->visited_ = true;
+    if (visiting->producer_edges.size() > 0 &&
+        std::any_of(
+            visiting->producer_edges.begin(),
+            visiting->producer_edges.end(),
+            [&](SegmentedEdge* dep) { return !dep->from->visited_; })) {
```

Comment:
```suggestion
    if (std::any_of(
            visiting->producer_edges.begin(),
            visiting->producer_edges.end(),
            [&](SegmentedEdge* dep) { return !dep->from->visited_; })) {
```
According to https://en.cppreference.com/w/cpp/algorithm/all_any_none_of, any_of returns false on an empty range.